### PR TITLE
Feat/audio toggle

### PR DESCRIPTION
**Pull request by SebastianLjungman**
Working volume controls, but after clicking the volume slider to change volume, and leaving the volume slider, keyboard shortcuts stop working until part of the window is clicked. **No issues** when the controls **aren't** conditionally render (always displayed).

The "progress" part of the volume slider should perhaps ideally by colored for clarity (i.e. 50% volume colors half the slider blue), but this couldn't be accomplished through CSS only on Webkit browsers. Important or not?

### CODE REVIEW

File: packages/core/src/WebPlayer.ts [hunk 1]
```diff
@@ -143,6 +143,35 @@ export default class WebPlayer extends EventEmitter {
     }
   }
 
+  changeVolume(change) {
```
Comment (bwallberg) [inline packages/core/src/WebPlayer.ts:146]: The common convention is to name things `set` not `change`, so `setVolume()`

Why do we have two methods for this?

File: packages/core/src/WebPlayer.ts [hunk 2]
```diff
@@ -143,6 +143,35 @@ export default class WebPlayer extends EventEmitter {
...
+    if (change > 0 && this.tech.volume < 1) {
```
Comment (bwallberg) [inline packages/core/src/WebPlayer.ts:147]: the code here does the exact same thing, could just be in a single if statement.

File: packages/core/src/WebPlayer.ts [hunk 3]
```diff
@@ -143,6 +143,35 @@ export default class WebPlayer extends EventEmitter {
...
+        this.tech.volume = (this.tech.volume + change).toFixed(2);
+    }
+    else if (change < 0 && this.tech.volume > 0) {
+      this.tech.volume = (this.tech.volume + change).toFixed(2);
+    }
+
+    if (this.tech.volume === 0) {
+      this.mute();
```
Comment (bwallberg) [inline packages/core/src/WebPlayer.ts:155]: is this really needed?

File: packages/eyevinn-skin/src/components/buttons/volume/VolumeControls.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { h } from 'preact';
+import style from './volume.module.css';
+import { useState } from 'preact/hooks';
+
+import VolumeButton from './VolumeButton';
+import VolumeSlider from './VolumeSlider';
+
+// function toggleMuteAndSlider() {
```
Comment (bwallberg) [inline packages/eyevinn-skin/src/components/buttons/volume/VolumeControls.tsx:8]: remove un-used code

File: packages/core/src/WebPlayer.ts [hunk 5]
```diff
@@ -143,6 +143,35 @@ export default class WebPlayer extends EventEmitter {
     }
   }
 
+  changeVolume(change) {
+    if (change > 0 && this.tech.volume < 1) {
+        this.tech.volume = (this.tech.volume + change).toFixed(2);
```
[the remark this answers is not in the document]
Comment (SebastianLjungman) [reply packages/core/src/WebPlayer.ts:148]: Done; I round the number in the volume setter in BaseTech instead.

File: packages/core/src/WebPlayer.ts (hunk 3 above)
Comment (SebastianLjungman) [reply packages/core/src/WebPlayer.ts]: See revised logic in latest commit.

**Comment by birme** [comment]
I've tried it out now and from a functionality perspective it looks ok for me. If the comments from Benjamin has been addressed I think this PR is ready.

File: packages/core/src/WebPlayer.ts [hunk 6]
```diff
@@ -143,6 +143,34 @@ export default class WebPlayer extends EventEmitter {
...
+    }
+    else if (change > 0 && this.tech.volume < 1) {
+      this.tech.volume = this.tech.volume + change;
+    } 
+    else if (change < 0 && this.tech.volume > 0) {
+      if (this.tech.volume + change === 0) {
+        this.mute();
+      }
+      else {
+      this.tech.volume = this.tech.volume + change;
```
Comment (birme) [inline packages/core/src/WebPlayer.ts:158]: please indent

File: packages/core/src/WebPlayer.ts [hunk 7]
```diff
@@ -143,6 +143,34 @@ export default class WebPlayer extends EventEmitter {
...
+    }
+  }
+
+  setVolumeByPercentage(percentage: number) {
+    if (percentage === 0) {
+      this.mute();
+      this.tech.volume = 0.1;
+    }
+    else {
+    this.tech.volume = percentage / 100;
```
Comment (birme) [inline packages/core/src/WebPlayer.ts:169]: please indent

File: packages/core/src/WebPlayer.ts (hunk 1 above)
Comment (SebastianLjungman) [reply packages/core/src/WebPlayer.ts]: Addressed in new commit: Now using optional parameters in setVolume

File: packages/core/src/WebPlayer.ts [hunk 8]
```diff
@@ -147,6 +147,41 @@ export default class WebPlayer extends EventEmitter {
     }
   }
 
+  setVolume({ 
+    change, 
+    percentage 
+  }: {
```
Comment (birme) [inline packages/core/src/WebPlayer.ts:153]: From a programming logic perspective I would have seen something more simplified. In the line of this

```
let newVolume = change ? this.tech.volume + change : percentage / 100;
if (newVolume > 1) {
  newVolume = 1;
}
if (newVolume < 0) {
  newVolume = 0;
}
this.tech.volume = newVolume;
this.tech.volume > 0 ? this.isMuted && this.unmute() : this.mute();
```

**Review by birme — CHANGES_REQUESTED** [review]
I think we will have to go for the <div> route as it seems difficult to get it to work that good with input[range]. Also as current mute/unmute icon is on the right I think we should keep it there and instead have a vertical slider. And we need to make sure that everything fits in a mobile display size context and when embedded.

I also suggest to split this up in two PR:s. One to add the function to set volume in core and another one for the skin. Enabling the set volume functionality in core. I will create a separate ticket for that.

**Pull request closed without merging**